### ISSUE TITLE
Matlab: bug with matchit plugin and "end" keyword.

### ISSUE THREAD

**Issue by gdupras** [vim/vim#8787]
When doing `%` on `if`, it may match `end` used to index a cell array instead of the `end` terminating the if-else block. For example:
```
x = {1, 2};
if x{end} == 2
    foo = x{1};
end
```
In the above, with the cursor on `if`, do `%` and the cursor lands on the `end` in `x{end}` instead of on the last `end`. I believe this is caused by an incorrect definition of the variable `b:match_words`, used by the matchit plugin.

I have contacted the maintainer Jake Wasserman listed in `runtime/ftplugin/matlab.vim` but got no reply. `.github/CODEOWNERS` has @cecamp has the owner of this file.

I'm using Vim 8.2.3346 on Windows 10.

**Comment by chrisbra** [comment]
I have a feeling that this line:

https://github.com/vim/vim/blob/a91871262670f9d3a4e5d1c5c6dbc02f85625f7d/runtime/ftplugin/matlab.vim#L18

has been added to handle this case. But I have a hard time deciphering the `s:conditionalEnd` item. I think it may work with the following even more ugly thing:
```patch
diff --git a/runtime/ftplugin/matlab.vim b/runtime/ftplugin/matlab.vim
index a1a282f19..fe6699566 100644
--- a/runtime/ftplugin/matlab.vim
+++ b/runtime/ftplugin/matlab.vim
@@ -15,7 +15,7 @@ let s:save_cpo = &cpo
 set cpo-=C

 if exists("loaded_matchit")
- let s:conditionalEnd = '\%(([^()]*\)\@!\<end\>\%([^()]*)\)\@!'
+ let s:conditionalEnd = '\%([{(][^(){}]*\)\@!\<end\>\%([^(){}]*[)}]\)\@!'
  let b:match_words=
    \ '\<\%(if\|switch\|for\|while\)\>:\<\%(elseif\|case\|break\|continue\|else\|otherwise\)\>:'.s:conditionalEnd.','.
    \ '\<function\>:\<return\>:\<endfunction\>'
```

But this is really just a guess.

**Comment by dkearns** [comment]
Shouldn't that first zero-width match be a look behind `\@<!` as well?

It might be easier just to match `end` at the beginning of a line or after a statement separator?

**Comment by gdupras** [comment]
Thank you @chrisbra for taking a look. The pattern you suggested does seem to
work correctly so far.

> It might be easier just to match end at the beginning of a line or after a
> statement separator?

That does sound easier but it doesn't quite work. But maybe I'm doing
something wrong though. I tried

    let s:conditionalEnd = '\(^\|;\)\s*\<end\>'

This works correctly except that the cursor can land on the white space or on
`;` in front of `end`. So I tried

    let s:conditionalEnd = '\(^\|;\)\s*\zs\<end\>'

and

    let s:conditionalEnd = '\(^\|;\)\s*\zs\<end\>\ze'

but when the cursor is on `end` and I do `%`, the cursor doesn't move and Vim
hangs for a few seconds.

`:help matchit` does mention that `\zs` "has not been thoroughly tested" so
maybe I ran into a matchit bug or I'm doing something wrong.

**Comment by dkearns** [comment]
> That does sound easier but it doesn't quite work. But maybe I'm doing
> something wrong though. I tried
> 
> ```
> let s:conditionalEnd = '\(^\|;\)\s*\<end\>'
> ```

Yes, `\zs` doesn't work well in matchit patterns for reasons I can't recall off the top of my head.  You can, however, use a positive lookbehind. E.g., 

`let s:conditionalEnd = '\%(\%(^\|;\)\s*\)\@<=\<end\>'`

**Comment by gdupras** [comment]
Thank you! This works perfectly so far. Maybe this could be added to Vim's runtime files? This file's maintainer seems to have gone missing.

 I've also added support for try/catch blocks.

```diff
diff --git a/C:/Program Files/Vim/vim82/ftplugin/matlab.vim b/after/ftplugin/matlab.vim
index a1a282f..ceadcb1 100644
--- a/C:/Program Files/Vim/vim82/ftplugin/matlab.vim
+++ b/after/ftplugin/matlab.vim
@@ -1,7 +1,7 @@
 " Vim filetype plugin file
 " Language:    matlab
 " Maintainer:  Jake Wasserman <jwasserman at gmail dot com>
-" Last Change:         2019 Sep 27
+" Last Change:         2021 Aug 30

 " Contributors:
 " Charles Campbell
@@ -15,9 +15,10 @@ let s:save_cpo = &cpo
 set cpo-=C

 if exists("loaded_matchit")
- let s:conditionalEnd = '\%(([^()]*\)\@!\<end\>\%([^()]*)\)\@!'
+ let s:conditionalEnd = '\%(\%(^\|;\)\s*\)\@<=end\>'
+
  let b:match_words=
-   \ '\<\%(if\|switch\|for\|while\)\>:\<\%(elseif\|case\|break\|continue\|else\|otherwise\)\>:'.s:conditionalEnd.','.
+   \ '\<\%(if\|switch\|for\|while\|try\)\>:\<\%(elseif\|case\|break\|continue\|else\|otherwise\|catch\)\>:'.s:conditionalEnd.','.       
    \ '\<function\>:\<return\>:\<endfunction\>'
  unlet s:conditionalEnd
 endif
```

**Comment by brammool** [comment]
I can include that.  But let's check with the maintainer one more time.

**Comment by dkearns** [comment]
Isn't `endfunction` an Octave only keyword?

**Comment by gdupras** [comment]
You're right. I suppose since Matlab and Octave are so similar they share the same file type plugin.

The syntax file does not define `endfunction` but it has a different maintainer.

Maybe a variable could be set to let the user choose which dialect they want.

**Comment by gdupras** [comment]
Fixed in 6c391a74fe90.
